kod/object/item/passitem/roomkeyc.kod

```diff
@@ -254,6 +254,16 @@ messages:
       propagate;
    }
 
+   Delete(bRemoveFromMasterKey=TRUE)
```

Comment:
Although Delete isn't a special function, it's a little jarring to see it take a parameter.  i.e. if you were deleting this from admin mode you certainly wouldn't expect a parameter.  Can you explain the case where bRemoveFromMasterKey=FALSE?  Why is it set to FALSE in OriginalDeleted?  (I think I just don't know how key copies work.)

Reply:
Done, and good catch.  The only times bRemoveFromMasterKey should be FALSE is if plCopies is going to be disposed of shortly afterward--in particular, when the room and master key are also being deleted.  (The change lock command, which also deletes the plCopies list, is fine either way so I've left it alone)

During deletion, the lookup chain this mess of a system needs to reach the master key breaks down, resulting in errors.  In light of this, I moved the #bRemoveFromMasterKey=FALSE into the copy-deleting loop in RoomKey's Delete function, where it makes sense.

Reply:
Something still doesn't feel right here.  The presence of any argument to Delete is fishy and unexpected.  Can Delete instead take no arguments, and attempt to delete itself from the master key, but do nothing if GetMasterKey returns $ ?  That would be a lot cleaner.

Reply:
That failed when In tried it before unfortunately.

However, this question prompted me to do some digging, and the reason it fails is because when the room expires and the process of cleaning it up begins, deletion of the key gets Posted.  When a copy is deleted, it attempts to look up its master key so it can remove itself from the list of copies, but this lookup fails since there's no longer a valid room and/or renter.  I found a clean way around this by just deleting all the key copies first thing, before any of the room cleanup begins.

As a bonus, this approach also allowed me to replace an adjacent Send(&RoomKey ...) function call.